pkg/cap/pubsub/client.go

```diff
@@ -52,28 +65,31 @@ func (t Topic) Publish(ctx context.Context, b []byte) error {
 	return err
 }
 
-func (t Topic) Subscribe(ctx context.Context, ch chan<- []byte) (cancel func(), err error) {
-	hc := api.Topic_Handler_ServerToClient(handler{
+func (t Topic) Subscribe(ctx context.Context, ch chan<- []byte) (release capnp.ReleaseFunc, err error) {
+	h := api.Topic_Handler_ServerToClient(handler{
 		ms:      ch,
 		release: t.AddRef().Release,
 	}, &server.Policy{
 		MaxConcurrentCalls: cap(ch),
 	})
-	defer hc.Release() // ensure topic ref is released if we return an error
+
+	// Limit the number of in-flight requests to half of the channel's capacity.
+	// This prevents the server from running out of memory due to the unbounded
+	// send-queue in capnp.
+	//
+	// TODO:  make this configurable.
+	h.Client.SetFlowLimiter(newFlowLimiter(cap(ch) / 2))
 
 	f, release := api.Topic(t).Subscribe(ctx, func(ps api.Topic_subscribe_Params) error {
-		return ps.SetHandler(hc.AddRef())
+		return ps.SetHandler(h)
 	})
 	defer release()
 
-	select {
-	case <-f.Done():
-		_, err = f.Struct()
-	case <-ctx.Done():
-		err = ctx.Err()
+	if _, err = f.Struct(); err == nil {
```

Comment:
Is it safe to get rid of `case <-f.Done():` and `case <-ctx.Done():` ?

Reply:
Yes, the `Struct()` call includes that select-case in its implementation, so it ended up being redundant.